### DISCUSSION TITLE
Migrate to Flux v2

### DISCUSSION THREAD

**Discussion by stefanprodan** [fluxcd/flux2#413] [Show and tell]
## Migrate from Flux v1

If you are using Flux v1, please consider retooling your CD pipelines with flux2.

Starting with flux2 v0.8, the default controllers and their APIs are considered stable and **production ready**, please see the [roadmap](https://toolkit.fluxcd.io/roadmap/) for a more details.

To find out what differences are between v1 and v2 please read the [FAQ](https://toolkit.fluxcd.io/faq/#flux-v1-vs-v2-questions).

### Flux guides

- [Migrate from Flux v1](https://toolkit.fluxcd.io/guides/flux-v1-migration/)
- [Migrate from `.flux.yaml` and kustomize](https://toolkit.fluxcd.io/guides/flux-v1-migration/)
- [Migrate from Flux v1 automated container image updates](https://toolkit.fluxcd.io/guides/flux-v1-automation-migration/)
- [How to manage multi-tenant clusters with Flux v2](https://github.com/fluxcd/flux2-multi-tenancy)

### Helm guides

- [Migrate from Helm Operator to Flux v2](https://toolkit.fluxcd.io/guides/helm-operator-migration/)
- [How to structure your HelmReleases](https://github.com/fluxcd/flux2-kustomize-helm-example)

**Comment by fabito** [comment]
Utility snippet for removing the finalizers:

```bash
{
for resource in kustomization gitrepository helmrelease helmrepository; do
    for item in $(kubectl get ${resource} -n gotk-system -o jsonpath='{.items[*].metadata.name}'); do
        kubectl -n gotk-system patch $resource $item -p '{"metadata":{"finalizers":null}}' --type=merge
    done
done
}
```

**Comment by dholbach** [comment]
As the big area of "image update functionality" is nearing completion (80% done, 2021-02-03), the [image update migration guide](https://github.com/fluxcd/flux2/pull/806) is looking for feedback and early testing. If you are interested in this area, please go check it out.

**Comment by dholbach** [comment]
We are getting closer to a request-for-testing period, before we call it GA.

What are relevant steps we want to take in the next weeks?

1. Write a blog post about testing the migration docs and giving feedback?
2. Link recent talks/presentations from guides and blog post?
3. Create a general migration page in the flux2 docs?
4. What else?

**Reply by pratikbin** [reply]
do live migration from v1 to v2 on youtube or zoom call maybe

**Reply by dholbach** [reply]
https://www.meetup.com/GitOps-Community/events/276539791/ cc @staceypotter and @stealthybox

**Reply by staceypotter** [reply]
👋  Hot off the press is this 📺  video by Leigh (@stealthybox) on Migrating from Flux v1 -> v2: https://youtu.be/vwvTwLQhXVI

**Comment by edobry** [comment]
The "Migrate from .flux.yaml and kustomize" doc looks great! Are there any plans to write one for a generic generator "flux.yaml", for example if using a custom command to generate manifests?

**Reply by edobry** [reply]
ah I see now in the roadmap:
```
Non-Goals
- Migrate users that are using Flux to run custom scripts with flux.yaml 
```
this is unfortunate! will this usecase not be supported with Flux v2?

**Reply by hiddeco** [reply]
That is indeed exactly what this means, for an explanation on why we decided to drop the support for this, see: https://toolkit.fluxcd.io/guides/faq-migration/#are-there-any-breaking-changes

**Reply by edobry** [reply]
I see, thanks for the response! So the only path to using Flux v2 if we have a custom manifest generation script is to build our own controller, similar to the Kustomization Controller?

**Reply by yebyen** [reply]
I am preparing a suite of example GitHub Actions that do manifest generation to cover several well-known `.flux.yaml` use cases, it has just landed in a draft PR but is not really ready for review just yet.

The decision at this point is, if you need some manifest generation that is above and beyond what Kustomize and envsubst provide, it should be done on a CI platform and published through a git branch. You can use any custom binary that you want in a CI job, we will publish further additional examples for different CI platforms as time and broadly expressed interest dictates.

**Reply by edobry** [reply]
The CI approach makes sense, its what I'm currently pursuing as well; the main question is how to support multiple different environment branches. Presumably the CI job will have to commit generated manifests to some branch, and Flux configured to pick them up; but if there are MULTIPLE environment branches it seems unclear how to set this up. Thanks for the response!

**Reply by kingdonb** [reply]
Indeed it is more complicated that way, and most users probably will have multiple environments so this is not a scenario that we can ignore. The guide is WIP in https://github.com/fluxcd/flux2/pull/1083/files

There are several note/hint callouts that are intended to lead you in the direction of a better solution for this, but I'm afraid it's still not very clear and the PR needs more work to merge.

**Reply by kingdonb** [reply]
The section at the bottom is titled "Adapting for Flux v2" and has the examples that are closest to what Flux v2 canonically recommends; the examples in the body of the guide are more focused on making things work as closely to how they work in Flux v1. There is an example that uses `release` branches, there might be a general solution for a fixed list of environments here but it is not documented in the examples yet. (Thanks for the feedback!)

> You probably want a `GitRepository` Source (and thus also branch) per environment,

(edit: I wrote this statement above out of reflex, but it is questionable. I do not think environment branches are the recommended way to handle multi-cluster in Flux v2.)

... and you can use either manifest gen techniques to write the commit to deploy into each branch like the guide describes, or use Flux v2 `ImagePolicy` automation. I think the suggestion is to use an `ImagePolicy` per environment rather than a branch per environment, and keep all the cluster environments in the same branch. It's not 100% clear to me how you could use a different ImagePolicy per environment branch.

I think this is where Kustomization bases and resource references are meant to be used to solve the problem in the most modern/updated way, without environment branches.

**Comment by sanjvij** [comment]
Hi . I am having a hard time in setting up flux on my Mac using bootstrap. when I try to run bootstrap command I get a health check error as below. Am I okay to ignore this ? 

◎ waiting for Kustomization "flux-system/flux-system" to be reconciled
✗ failed to download artifact, error: GET http://source-controller.flux-system.svc.cluster.local./gitrepository/flux-system/flux-system/3f5e6a9863de3306a0ae2f9a69cf0fa62d558e0e.tar.gz giving up after 10 attempt(s)
► confirming components are healthy
✔ source-controller: deployment ready
✔ kustomize-controller: deployment ready
✔ helm-controller: deployment ready
✔ notification-controller: deployment ready
✔ all components are healthy
✗ bootstrap failed with 1 health check failure(s)

**Reply by stefanprodan** [reply]
Sorted out here https://github.com/fluxcd/flux2/issues/1265 It was a transient error due to slow network.

**Comment by leandro-loos** [comment]
Hi... I use flux 1 in a very basic way, I have a deployment yaml file with the image provided by you guys on dockerhub  and with some `args` on said deployment file I configure it to look for changes on a specific branch within a specific repo and apply them, and that is enough to make me happy and to be able to deploy it from anywhere without needing to do anything locally apart from `kubectl apply ....`

Is it possible to have something like that working with flux2? Any links to deployment files I could use as an example?

**Reply by yebyen** [reply]
The bootstrap guide shows how to do this, it is more moving parts than before, but what you get with bootstrap that you don't get with `flux install` is a straightforward and well documented upgrade path.

What you want is a single GitRepository and a single Kustomization that applies whatever yaml file or files you push in the git tree. Bootstrap provides the example, it produces two files, one is called "gotk-sync"

We recommend the bootstrap guide, or Getting Started guide, for Flux beginners because it teaches these important concepts that you can't use Flux without. However, as more complicated setups are explained in follow-on links that appear at the end of the Getting Started guide, you don't need to follow them if what you want is just what you described. You can simplify and pare down instead of scaling up.

The simplest Flux v2 installation is just `flux install` (no bootstrap) with just a single GitRepository and Kustomization as Bootstrap creates. If you need help with this, you are welcome to join us on #flux in CNCF Slack and we will be glad to help you work through any issues. 👍

**Comment by ekosov-form3** [comment]
Thanks for all the migration guides. They all look great! :+1: 
However, I'd like to make sure that my understanding is correct. It appears to me that only "big bang" (everything at once) kind of migration is supported, as there's "stop/uninstall flux v1" step in there. What we're trying to achieve though is a gradual migration, when some subset of resources is migrated from v1 to v2, while having both v1 and v2 running in parallel. Is that possible w/o disruption to workloads?

**Reply by ekosov-form3** [reply]
I'll just add a couple of points to clarify our motivation and issues.

We're managing quite a bit (100s) of resources by Flux and we're not quite comfortable migrating all of them to Flux v2 at once, esp. in production. Hence we're trying to run both v1 and v2 in parallel and gradually migrate manifests from v1 to v2. The issue we're seeing with this approach though is that (sometimes) flux v1 tries to delete resources which are being taken over by flux v2, which causes disruption.

The process is something like this
 - scale flux v1 down
 - migrate manifests (in code) from v1 to v2
 - let flux v2 reconcilers do the thing (take existing resources over)
 - scale flux v1 back up

When flux v1 starts back up and discovers that some of the manifests are not there, while resources are still present, it removes them. 

As a workaround, we're considering removing the `fluxcd.io/sync-gc-mark` label, so flux v1 looses track of the resource. There might be a better solution to this. If not, perhaps, it could be done by flux v2 (kustomize-controller?) automatically, when running in e.g. "migration mode"?

Perhaps, it's much simpler and I'm just missing some docs.
Appreciate any suggestions on this :pray:

**Reply by kingdonb** [reply]
The `fluxcd.io/ignore: "true"` annotation is only honored by Flux v1, so if you put it on any resources that you are wary of Flux v1 "managing" into the garbage bin at the end of its life with respect to the resource, it should be enough to prevent that from happening.

You could use `commonAnnotations` in a Kustomize overlay to apply that label across the board in Flux v2 on a per-resource basis, but that seems a bit awkward and I really wouldn't expect you to go on using that for very long, as we didn't intend for folks to go on using Flux v1 for very long once they've started the migration. The goal from our perspective is to get users to stop using Flux v1 so we can sunset it. As long as there are active users we can't really do that without putting users like you out in the cold.

If you are using deprecated features like `.flux.yaml` or Helm plugins, and simply can't solve those use cases any other way, please make sure your issue is documented for Flux v2; we have asked that all user issues preventing folks from migrating are documented as issues so that we can track them and be sure we're not leaving anyone behind when GA is finally declared on Flux v2. So if there are specific reasons (besides inertia and massiveness, with the effort you are taking to manually verify each resource as it gets migrated) please make sure that they get documented.

Lots of effort was placed on the migration documents, but there could still be gaps. If I was to suggest where issues are most likely to come from, it will be from incompletely migrating, running both versions side-by-side, which is a configuration I personally haven't tested very well together or very long, or well enough to know where the worst edge cases are likely to be encountered. Not to mince words, I would recommend solving all issues in a staging environment, then completing the migration at once when all known issues are solved. While possible, it's strictly not recommended to run both Flux v1 and Helm Operator + Flux v2 and Helm Controller all together, and I wouldn't really recommend doing so in production for very long, if even at all.

I've tried to focus more on gaining the relevant experience with Flux v2 so that I can support the migration for users of Flux v2 and advertise all of the relevant features there, with as little emphasis as possible on the old versions. We have tried to make sure the migration is smooth and that all important use cases are addressed by it, so folks can get all the way on Flux v2. As a consequence, there is simply less attention on issues related to running Flux v1 even from myself, the lead primarily responsible for Flux v1 maintenance. Transparently, my most important goal is to get everyone off the boat first.

**Reply by ekosov-form3** [reply]
@kingdonb thanks for your quick response, much appreciated. :raised_hands: 
I actually like the idea of temporarily adding the "ignore" annotation to all the of the "already migrated" resources. That should do the trick.
I'm going to give this a try and get back to you in case of any problems.
Thank you.
